### PR TITLE
[BEAM-6619] [BEAM-6593] Add gcsio integration tests to postcommit

### PR DESCRIPTION
**Pull request by Juta**
This is is part of a series of PRs with goal to make Apache Beam PY3 compatible. The proposal with the outlined approach has been documented here: [https://s.apache.org/beam-python-3](https://s.apache.org/beam-python-3).

This PR adds a more integration tests to the postcommit jobs on direct and dataflow runners.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python3 PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by tvalentyn** [comment]
Run Python PostCommit

File: sdks/python/test-suites/dataflow/py3/build.gradle [hunk 1]
```diff
@@ -34,15 +34,25 @@ def basicTestOpts = [
...
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_rewrite_token",
+        "apache_beam.examples.complete.game.user_score_it_test:UserScoreIT.test_user_score_it",
     ]
     def testOpts = basicTestOpts + ["--tests=${tests.join(',')}"]
     def cmdArgs = project.mapToArgString([
         "test_opts": testOpts,
         "sdk_location": "${project.buildDir}/apache-beam.tar.gz"
     ])
     exec {
+      environment 'PYTHONHASHSEED', '0'
```
Comment (tvalentyn) [inline sdks/python/test-suites/dataflow/py3/build.gradle]: Let's comment why this is needed.

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by tvalentyn** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PostCommit

**Comment by Juta** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ def tearDown(self):
   @attr('IT')
   def test_streaming_wordcount_it(self):
     # Build expected dataset.
-    expected_msg = [('%d: 1' % num) for num in range(DEFAULT_INPUT_NUMBERS)]
+    expected_msg = [(b'%d: 1' % num) for num in range(DEFAULT_INPUT_NUMBERS)]
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: I think we should instead decode `bytes` that we receive from PubSub in https://github.com/apache/beam/blob/79f8e34fc75c59297c0c7f07ce53325ce96bb5cf/sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py#L106.

The matcher expects strings:
https://github.com/apache/beam/blob/79f8e34fc75c59297c0c7f07ce53325ce96bb5cf/sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py#L58

PubSub client returns bytes: https://github.com/apache/beam/blob/a5ed104aeac91781492523808fa4d3df6f2e55f7/sdks/python/apache_beam/io/gcp/pubsub.py#L60

Given this is a test code we can assume a 'utf-8' encoding

File: sdks/python/test-suites/dataflow/py3/build.gradle [hunk 3]
```diff
@@ -34,8 +34,16 @@ def basicTestOpts = [
...
         "apache_beam.examples.cookbook.bigquery_tornadoes_it_test:BigqueryTornadoesIT.test_bigquery_tornadoes_it",
+        "apache_beam.examples.streaming_wordcount_it_test:StreamingWordCountIT.test_streaming_wordcount_it",
+        "apache_beam.examples.wordcount_it_test:WordCountIT.test_wordcount_it",
+        "apache_beam.examples.wordcount_it_test:WordCountIT.test_wordcount_fnapi_it",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_batch",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_batch_kms",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_batch_rewrite_token",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_kms",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_rewrite_token",
```
Comment (markflyhigh) [inline sdks/python/test-suites/dataflow/py3/build.gradle]: Please make sure `"--process-timeout=4500"` above is a reasonable amount. This is a [nose flag](https://nose.readthedocs.io/en/latest/man.html#cmdoption-process-timeout) that stands for total timeout of running all tests together.

File: sdks/python/test-suites/direct/py3/build.gradle [hunk 4]
```diff
@@ -28,8 +28,16 @@ task postCommitIT(dependsOn: 'installGcpTest') {
...
+        "apache_beam.examples.wordcount_it_test:WordCountIT.test_wordcount_fnapi_it",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_batch",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_batch_kms",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_batch_rewrite_token",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_kms",
+        "apache_beam.io.gcp.gcsio_integration_test:GcsIOIntegrationTest.test_copy_rewrite_token",
     ]
     def testOpts = [
         "--tests=${batchTests.join(',')}",
```
Comment (markflyhigh) [inline sdks/python/test-suites/direct/py3/build.gradle]: You can set parallel run by adding [`--processes`](https://nose.readthedocs.io/en/latest/man.html#cmdoption-processes) and [`--process-timeout`](https://nose.readthedocs.io/en/latest/man.html#cmdoption-process-timeout) here if running above tests in sequence take too much time. An example is in [here](https://github.com/apache/beam/blob/master/sdks/python/test-suites/dataflow/py3/build.gradle#L28).

**Comment by Juta** [comment]
Run Python PostCommit

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py (hunk 2 above)
Comment (Juta) [reply sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: Ok thank you! I changed it

**Comment by tvalentyn** [comment]
LGTM, thanks, @Juta !

**Comment by charlesccychen** [comment]
Thanks, this LGTM.

**Pull request merged**